### ISSUE TITLE
kOS replaces "windows new line" (CR+LF) with CR+CR+LF when blocks of code is pasted directly into the kOS code editor.

### ISSUE THREAD

**Issue by Zyt3x** [KSP-KOS/KOS#639]
Tried searching for an already existing issue, but the nearest I got was #206 .

I have this code in Notepad++:
```
SAS on.CRLF
RCS off.
```
Due to an unrelated bug (hopefully), I have to copy+paste the code into the in-game code editor and save it before trying to run the script for the first time. After saving the code it looks like this:
```
SAS on.CR
CRLF
RCS off.
```

Running Windows 7 64-bit, 32-bit KSP. Code encoded as UTF-8.

**Comment by Dunbaratu** [comment]
That's because the editor thinks you're pressing all those keys and typing *very fast* when you pasted in.  It thought you did the following:
User typed "S".
User typed "A".
User typed "S".
User typed " ".
User typed "o".
User typed "n".
User typed ".".
User typed 0x13 key (return, or control-M).
User typed 0x10 key (linefeed, or control-J).

So when you "typed" linefeed, it translated that into CRLF "for you", thinking it had to do the \\n to \\r\\n mapping that so often has to be done in Windows when using a C or C-like language.

Normally if you're typing, you don't end up typing the LF separately, so it does the conversion of the one character into two.

Not sure about how to fix it.

**Comment by Zyt3x** [comment]
I see. It's not possible to check if the last key entered was 13..?

**Comment by Dunbaratu** [comment]
What makes that a pain is that if you type return return return at the keyboard, it needs to see that as being 3 lines, not one.

Also, this behavior isn't entirely under our control anyway.  We're just using a built-in Unity textarea widget.  Unity is dictating how cut-and-paste works in that widget, not us.  I'm not sure how to fix it other than to be able to detect upon reading the result of the widget what Unity did to mess up its contents and undo that messing up of them.

**Comment by Zyt3x** [comment]
Figured it was Unity, there's no way anyone sane would program that in like that in intentionally :P 
It's a shame it's not an easy fix.. I'm guessing you can't add in a paste button either?

**Comment by ZiwKerman** [comment]
@Zyt3x But not just use normal text editors to edit `"yourfile.ks"` in `Ships\Script` and then `copy "yourfile.ks" to desiredVolume.` or even straight out `switch to 0. run yourfile.`?

**Comment by Zyt3x** [comment]
Haven't tried "copy "yourfile.ks" to desiredVolume.", am currently doing "switch to 0. run yourfile.", but as stated before; I ran into problems when trying to run the script for the first time without loading+saving using the kOS editor

**Comment by ZiwKerman** [comment]
It may be a problem with permissions on `Ships\Script` folder, have checked there?

**Comment by Dunbaratu** [comment]
Yeah I'm actually more interested in this weird need to load the file into the kOS editor first.  As annoying as the behavior of the unity text widget is when it comes to cut and paste, I'm more worried about the claim that you *have* to do it that way.  That indicates something seriously broken elsewhere.  I'm more concerned with hearing about THAT and trying to fix it.

**Comment by Zyt3x** [comment]
Filed new issue #644 for the kOS editor shenanigans.
